src/signing.rs

```diff
@@ -817,55 +799,196 @@ impl ExpandedSecretKey {
+    hash_prefix: [u8; 32],
+    prehashed_message: D,
+    verifying_key: &VerifyingKey,
+    context: Option<&'a [u8]>,
+) -> Result<Signature, SignatureError>
+where
+    D: Digest<OutputSize = U64>,
+{
+    let mut h: Sha512;
+    let mut prehash: [u8; 64] = [0u8; 64];
+
+    let ctx: &[u8] = context.unwrap_or(b""); // By default, the context is an empty string.
+
+    if ctx.len() > 255 {
+        return Err(SignatureError::from(InternalError::PrehashedContextLength));
+    }
+
+    let ctx_len: u8 = ctx.len() as u8;
+
+    // Get the result of the pre-hashed message.
+    prehash.copy_from_slice(prehashed_message.finalize().as_slice());
+
+    // This is the dumbest, ten-years-late, non-admission of fucking up the
+    // domain separation I have ever seen.  Why am I still required to put
+    // the upper half "prefix" of the hashed "secret key" in here?  Why
+    // can't the user just supply their own nonce and decide for themselves
+    // whether or not they want a deterministic signature scheme?  Why does
+    // the message go into what's ostensibly the signature domain separation
+    // hash?  Why wasn't there always a way to provide a context string?
+    //
+    // ...
+    //
+    // This is a really fucking stupid bandaid, and the damned scheme is
+    // still bleeding from malleability, for fuck's sake.
+    h = Sha512::new()
+        .chain_update(b"SigEd25519 no Ed25519 collisions")
+        .chain_update([1]) // Ed25519ph
+        .chain_update([ctx_len])
+        .chain_update(ctx)
+        .chain_update(hash_prefix)
+        .chain_update(&prehash[..]);
+
+    let r = Scalar::from_hash(h);
+    let R: CompressedEdwardsY = EdwardsPoint::mul_base(&r).compress();
+
+    h = Sha512::new()
+        .chain_update(b"SigEd25519 no Ed25519 collisions")
+        .chain_update([1]) // Ed25519ph
+        .chain_update([ctx_len])
+        .chain_update(ctx)
+        .chain_update(R.as_bytes())
+        .chain_update(verifying_key.as_bytes())
+        .chain_update(&prehash[..]);
+
+    let k = Scalar::from_hash(h);
+    let s: Scalar = (k * scalar) + r;
+
+    Ok(InternalSignature { R, s }.into())
+}
+
+//
+// Exports
+//
+
+/// The plain, non-prehashed, signing function for Ed25519.
+///
+/// **Unsafe:** Do NOT use this function unless you absolutely must. Misuse of this function can
+/// expose your private key. See [here](https://github.com/MystenLabs/ed25519-unsafe-libs) for more
+/// details on this attack.
+///
+/// # Inputs
+///
+/// * `scalar` is the secret scalar of the signing key
+/// * `hash_prefix` is the domain separator that, along with the message itself, is used to
+///   deterministically generate the `R` part of the signature.
+///
+/// `scalar` and `hash_prefix` are usually selected such that `scalar || hash_prefix = H(sk)` where
+/// `sk` is the signing key
+#[cfg(feature = "raw-sign")]
+fn raw_sign(
+    scalar: Scalar,
+    hash_prefix: [u8; 32],
+    message: &[u8],
+    verifying_key: &VerifyingKey,
```

Comment:
One thing that makes this API a bit unpleasant, but certainly not unworkable, is that we don't have a simple method to construct a `VerifyingKey` from a `Scalar`.

`VerifyingKey::clamp_and_mul_base()`, which is used by `ExpandedSecretKey`, is private as well, so we have no other choice than going through `VerifyingKey::from_bytes()`.

Reply:
Would `impl From<EdwardsPoint> for VerifyingKey` be sufficient?